storm-core/src/jvm/org/apache/storm/StormSubmitter.java

```diff
@@ -238,39 +239,24 @@ public static void submitTopologyAs(String name, Map stormConf, StormTopology to
                 }
 
                 // Dependency uploading only makes sense for distributed mode
-                List<String> jarsBlobKeys;
+                List<String> jarsBlobKeys = Collections.emptyList();
                 List<String> artifactsBlobKeys;
 
                 DependencyUploader uploader = new DependencyUploader();
                 try {
                     uploader.init();
                     jarsBlobKeys = uploadDependencyJarsToBlobStore(uploader);
                     artifactsBlobKeys = uploadDependencyArtifactsToBlobStore(uploader);
+                    setDependencyBlobsToTopology(topology, jarsBlobKeys, artifactsBlobKeys);
+                    submitTopologyInDistributeMode(name, topology, opts, progressListener, asUser, conf, serConf);
+                } catch (Throwable e) {
+                    // remove uploaded jars blobs, not artifacts since they're shared across the cluster
+                    uploader.deleteBlobs(jarsBlobKeys);
```

Comment:
@satishd 
In fact, submitTopologyWithOpts() can also throw AuthorizationException and org.apache.thrift.TException, too.

Let's back to flow of logic.
Please note that we didn't catch Throwable (some known Exceptions and also RuntimeException) and just pass to caller. And there's also no return for Nimbus.Client.submitTopology(), which indicates that unless we get TException or other Exceptions it should be treated as succeed. (If not we can't guarantee topology deploy is succeed.)

I think this assumption is right for now, and assumption should be right unless we add some operations after calling submitTopologyInDistributeMode() or adding some operations after logging "Finished submitting topology".